### PR TITLE
Fog of war implementation

### PR DESCRIPTION
**Pull request by alkorolyov**
I somehow messed up the previous PL in such way that it is easier to create a new one.

I added a "Fog of war" button, which is seen only in spectator mode.
It could be clicked to to toggle on/off fog of war.  Now all unexplored tiles looked good. Answering your previous question on how to switch players view, we can toggle off fog of war, and then select new player.

Please find attached a screenshot on how it looks.

fog_1

### CODE REVIEW

**Comment by alkorolyov** [comment]
@yairm210 
Finished fog of war. Should be ready for merger now.

File: core/src/com/unciv/ui/worldscreen/WorldScreen.kt [hunk 1]
```diff
@@ -460,6 +477,20 @@ class WorldScreen(val viewingCiv:CivilizationInfo) : CameraStageBaseScreen() {
...
     }
 
+    private fun createFogOfWarButton(): TextButton {
+        val fogOfWarButton = TextButton("Fog of War", skin)
+        fogOfWarButton.label.setFontSize(30)
+        fogOfWarButton.labelCell.pad(10f)
+        fogOfWarButton.pack()
+        fogOfWarButton.onClick {
+            if (fogOfWar) fogOfWar = false
+            else fogOfWar = true
```
Comment (yairm210) [inline core/src/com/unciv/ui/worldscreen/WorldScreen.kt:487]: fogOfWar = !fogOfWar

File: core/src/com/unciv/ui/worldscreen/WorldScreen.kt [hunk 2]
```diff
@@ -460,6 +477,20 @@ class WorldScreen(val viewingCiv:CivilizationInfo) : CameraStageBaseScreen() {
         else viewingCiv
     }
 
+    private fun createFogOfWarButton(): TextButton {
+        val fogOfWarButton = TextButton("Fog of War", skin)
```
Comment (yairm210) [inline core/src/com/unciv/ui/worldscreen/WorldScreen.kt:481]: .toTextButton(), for the translation. Also, add "Fog of War" as a line in the translations template.properties

File: core/src/com/unciv/ui/worldscreen/WorldScreen.kt (hunk 1 above)
Comment (alkorolyov) [reply core/src/com/unciv/ui/worldscreen/WorldScreen.kt:487]: Done!

File: core/src/com/unciv/ui/worldscreen/WorldScreen.kt (hunk 2 above)
Comment (alkorolyov) [reply core/src/com/unciv/ui/worldscreen/WorldScreen.kt:481]: ok, got it.

File: core/src/com/unciv/ui/tilegroups/WorldTileGroup.kt [hunk 3]
```diff
@@ -25,14 +25,17 @@ class WorldTileGroup(internal val worldScreen: WorldScreen, tileInfo: TileInfo,
...
             icons.addPopulationIcon()
-
-        val currentPlayerCiv = worldScreen.viewingCiv
-        if (UncivGame.Current.viewEntireMapForDebug
-                || currentPlayerCiv.exploredTiles.contains(tileInfo.position))
-            updateCityButton(city, tileIsViewable || UncivGame.Current.viewEntireMapForDebug) // needs to be before the update so the units will be above the city button
+        // update city buttons in explored tiles or entire map
+        if (showEntireMap || viewingCiv.exploredTiles.contains(tileInfo.position))
+            updateCityButton(city, tileIsViewable || showEntireMap) // needs to be before the update so the units will be above the city button
+        else if (UncivGame.Current.gameInfo.currentPlayerCiv.isSpectator())
```
Comment (yairm210) [inline core/src/com/unciv/ui/tilegroups/WorldTileGroup.kt:36]: I don't like the constant references to things trough the UncivGame.Current, that's to be avoided when possible.
I think this can be activated anyway, even if the current player isn't a spectator, no?

File: core/src/com/unciv/ui/tilegroups/TileGroup.kt [hunk 4]
```diff
@@ -285,11 +286,31 @@ open class TileGroup(var tileInfo: TileInfo, var tileSetStrings:TileSetStrings)
 
     fun isViewable(viewingCiv: CivilizationInfo) = showEntireMap
             || viewingCiv.viewableTiles.contains(tileInfo)
+            || !UncivGame.Current.worldScreen.fogOfWar
```
Comment (yairm210) [inline core/src/com/unciv/ui/tilegroups/TileGroup.kt:289]: Why is this needed? If fog of war is disabled, Isn't the viewingCiv the Speectator civ that can see everything anyway?
And if a check is needed then again to avoid the UncivGame.Current reference, I would check viewingCiv.isSpectator()

File: core/src/com/unciv/ui/tilegroups/WorldTileGroup.kt (hunk 3 above)
Comment (alkorolyov) [reply core/src/com/unciv/ui/tilegroups/WorldTileGroup.kt:36]: > I don't like the constant references to things trough the UncivGame.Current, that's to be avoided when possible.

Sorry, always forget about this. Will get rid of this static link.

> I think this can be activated anyway, even if the current player isn't a spectator, no?

Well, I added this additional check just for performance issues, because in the normal mode, we generally don't touch unexplored tiles until next turn. Only in spectator mode where we can change views we can access them on each update.

File: core/src/com/unciv/ui/tilegroups/TileGroup.kt (hunk 4 above)
Comment (alkorolyov) [reply core/src/com/unciv/ui/tilegroups/TileGroup.kt:289]: You are right, viewingCiv.isSpectator() will be enough here.

Also wanted to comment that is was really great decision of yours to add viewingCiv as an argument to updates. That really made fog of war implementation very easy.

File: android/assets/jsons/translations/template.properties [hunk 5]
```diff
@@ -509,8 +509,9 @@ Turn =
 turns = 
 turn = 
 Next unit = 
-Pick a policy = 
-Movement = 
+Fog of War =
+Pick a policy =
```
Comment (yairm210) [inline android/assets/jsons/translations/template.properties:513]: You removed the spaces after the = 0_0
Those are super important, also for the Fog of War

File: core/src/com/unciv/ui/worldscreen/WorldScreen.kt [hunk 6]
```diff
@@ -269,6 +277,12 @@ class WorldScreen(val viewingCiv:CivilizationInfo) : CameraStageBaseScreen() {
         }
     }
 
+    private fun getSpectatorCiv(): CivilizationInfo? {
```
Comment (yairm210) [inline core/src/com/unciv/ui/worldscreen/WorldScreen.kt:280]: I don't think this function is necessary
It seems that this function is only called when fog of war is false
And in that situation, we can just use viewingCiv, since the only time fogOfWar is false is when viewingCiv is spectator, no?

File: android/assets/jsons/translations/template.properties (hunk 5 above)
Comment (alkorolyov) [reply android/assets/jsons/translations/template.properties:513]: Sorry! -_-
I believe it is IDE somehow autocorrected it. Fixed it.

File: core/src/com/unciv/ui/worldscreen/WorldScreen.kt (hunk 6 above)
Comment (alkorolyov) [reply core/src/com/unciv/ui/worldscreen/WorldScreen.kt:280]: You are right fogOfWar could be false only in spectator mode. Thought could some meaningful fun name for clarity. But seems simple update(viewingCiv) would be enough.

Fixed it

**Comment by alkorolyov** [comment]
Added all fixes.
Actually there was some weird behavior from Android Studio, it "eats" spaces form the next line when editing, some sort of autocorrection I guess ...
